File: lib/lambda/rules/aws_cis_3_3.ts
import { cloudTrailRecord, uguisuRule, detection } from "../models";

// https://docs.aws.amazon.com/securityhub/latest/userguide/securityhub-cis-controls.html#cis-3.3-remediation

export class rule extends uguisuRule {
  constructor() {
    super({
      id: "aws_cis_3.3",
      title: "Usage of root account",
      description:
        "AWS CIS benchmark 3.3 recommend to ensure a log metric filter and alarm exist for usage of root account",
      severity: "medium",
    });
  }

  detect(record: cloudTrailRecord): detection | null {
    if (
      record.userIdentity !== undefined &&
      record.userIdentity.type === "Root" &&
      record.userIdentity.invokedBy === undefined &&
      record.eventType !== "AwsServiceEvent"
    ) {
      return {
        rule: this,
        event: record,
      };
    }

    return null;
  }
}

File: pkg/rules/aws_cis_3_3.go
package rules

import (
	"github.com/m-mizutani/uguisu/pkg/models"
)

// https://docs.aws.amazon.com/securityhub/latest/userguide/securityhub-cis-controls.html#cis-3.1-remediation
type awsCIS3_3 struct {
}

func newAwsCIS3_3() models.Rule {
	return &awsCIS3_3{}
}

func (x *awsCIS3_3) ID() string                { return "aws_cis_3.3" }
func (x *awsCIS3_3) Title() string             { return "Usage of root account" }
func (x *awsCIS3_3) Severity() models.Severity { return models.SeverityMedium }
func (x *awsCIS3_3) Description() string {
	return "AWS CIS benchmark 3.3 recommend to ensure a log metric filter and alarm exist for usage of root account"
}
func (x *awsCIS3_3) Match(record *models.CloudTrailRecord) bool {
	return record.UserIdentity.Type == "Root" &&
		record.UserIdentity.InvokedBy == nil &&
		record.EventType != "AwsServiceEvent"
}
